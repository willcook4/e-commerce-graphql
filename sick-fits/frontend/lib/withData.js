import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    // Local data
    clientState: {
      resolvers: {
        Query: {
          //
        },
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // read the cartOpen value from cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY
            })
            // Write the cart state to the opposite
            const data = {
              data: {
                cartOpen: !cartOpen
              }
            }
            cache.writeData(data)
            return data;
          },
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
