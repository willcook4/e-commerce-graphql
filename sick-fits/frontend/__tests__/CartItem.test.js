import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { MockedProvider } from 'react-apollo/test-utils'
import CartItem from '../components/CartItem'
import { fakeCartItem } from '../lib/testUtils'

describe('<CartItem />', () => {
  it('renders and matches the snapshot', async () => {
    const cartItem = fakeCartItem()
    const wrapper = mount(
      <MockedProvider>
        <CartItem cartItem={cartItem} />
      </MockedProvider>
    )
    
    const cartItemWrapper = wrapper.find('CartItem li') 
    expect(toJSON(cartItemWrapper)).toMatchSnapshot()
  })

it('handles an item that has been removed from the store (!cartItem.item)', () => {
    const cartItem = fakeCartItem({ item: null })
    const wrapper = mount(
      <MockedProvider>
        <CartItem cartItem={cartItem} />
      </MockedProvider>
    )

    const cartItemWrapper = wrapper.find('CartItem p')
    expect(cartItemWrapper.text()).toEqual('Sorry, this cart item has been removed')
  })
})