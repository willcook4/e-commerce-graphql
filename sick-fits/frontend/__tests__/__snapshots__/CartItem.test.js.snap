// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> renders and matches the snapshot 1`] = `
<li
  className="CartItem__CartItemStyles-sc-1rm9l7o-0 iguiiv"
>
  <img
    alt="dogs are best"
    src="dog-small.jpg"
    width="100"
  />
  <div
    className="cart-item-details"
  >
    <h3>
      dogs are best
    </h3>
    <p>
      $150
       - 
      <em>
        3
         × 
        $50
         each
      </em>
    </p>
  </div>
  <RemoveFromCart
    id="omg123"
  >
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": Object {
                "kind": "Name",
                "value": "removeFromCart",
              },
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "removeFromCart",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "ID",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 88,
            "start": 0,
          },
        }
      }
      optimisticResponse={
        Object {
          "__typename": "Mutation",
          "removeFromCart": Object {
            "__typename": "CartItem",
            "id": "omg123",
          },
        }
      }
      update={[Function]}
      variables={
        Object {
          "id": "omg123",
        }
      }
    >
      <RemoveFromCart__BigButton
        disabled={false}
        onClick={[Function]}
        title="Delete Item"
      >
        <button
          className="RemoveFromCart__BigButton-emvtd6-0 gPWNSn"
          disabled={false}
          onClick={[Function]}
          title="Delete Item"
        >
          ×
        </button>
      </RemoveFromCart__BigButton>
    </Mutation>
  </RemoveFromCart>
</li>
`;
